### PR TITLE
Yoon_updated

### PR DESCRIPTION
**Pull request by yoonbai**
Requesting code review before merging to main and aggregating intermediate (daily) NWB files into one merged NWB.
1. stimuli_interface.py: combined stimulus hashes and filenames to track stimulus ordering
2. added MWorks libraries to parse raw MWorks data files to organize stimulus events and constructing PSTHs
3. added quality control libraries 
4. implemented lab-specific pipeline that introduces a project configuration file (project_config.yaml) 
5. implemented lab-specific pipeline that saves intermediate NWB files from daily recordings

### CODE REVIEW

File: src/dicarlo_lab_to_nwb/conversion/convert_session.py
Comment (h-mayorquin) [inline src/dicarlo_lab_to_nwb/conversion/convert_session.py:78]: I left a comment here on where you have to do the changes.|

For reference, this is the place where the info is used in aggregation:
https://github.com/catalystneuro/dicarlo-lab-to-nwb/blob/a376254e97a498df9e0fbe18e49a7c66e1e2a20f/src/dicarlo_lab_to_nwb/conversion/aggregation.py#L313-L318

File: src/dicarlo_lab_to_nwb/conversion/psth.py
Comment (h-mayorquin) [inline src/dicarlo_lab_to_nwb/conversion/psth.py:105]: was this removed intentionally?

File: src/dicarlo_lab_to_nwb/conversion/stimuli_interface.py
Comment (h-mayorquin) [inline src/dicarlo_lab_to_nwb/conversion/stimuli_interface.py:212]: I don't see how you are using this one.

File: src/dicarlo_lab_to_nwb/conversion/stimuli_interface.py
Comment (h-mayorquin) [inline src/dicarlo_lab_to_nwb/conversion/stimuli_interface.py:215]: This part from down here should be reverted to use the iterator. This `SingleImageIterator` object was added to deal with memory problems.

File: src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_project.py
Comment (h-mayorquin) [inline src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_project.py:10]: I modified this here to the right import.

File: src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_project.py
Comment (h-mayorquin) [inline src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_project.py:63]: I added this variable for your control.

File: src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_project.py
Comment (h-mayorquin) [inline src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_project.py:123]: You can break at this level or parse inside, I think it is more transparent to parse as soon as possible but this is your call.

File: src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_project.py
Comment (h-mayorquin) [inline src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_project.py:128]: Nice. We could add an interface that handles all of this if you want. But this also gives more flexibility so probably not necessary.

File: src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_project.py
Comment (h-mayorquin) [inline src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_project.py:154]: Suggestion:

Abastract this away into a function that takes an in-memory nwb and calculates outputs the data frame `calculate_quality_metrics_from_nwb(nwbfile)`

That way, the code flow would be simpler and you can modify the function nwihtout modifying the running script.

File: src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_session.py
Comment (h-mayorquin) [inline src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_session.py:1]: We should erase the extra `conversion_session.py` file now that we have converged into one `convert_session.py`

File: src/dicarlo_lab_to_nwb/thresholding_pipeline/data_locator.py
Comment (h-mayorquin) [inline src/dicarlo_lab_to_nwb/thresholding_pipeline/data_locator.py:1]: Do we still need these guys now that your project takes file_paths directly?

File: src/dicarlo_lab_to_nwb/thresholding_pipeline/metadata.yaml
Comment (h-mayorquin) [inline src/dicarlo_lab_to_nwb/thresholding_pipeline/metadata.yaml:1]: Probably not needed, right?
Feel free to modify the metadata under conversion to suit your needs.

**Review by h-mayorquin — COMMENTED** [review]
OK, I pushed all the changes you can work only on this branch. That is:
1) I updated the `convert_session_nwb` to follow our discussion today in main and pushed it here.
2) I also pushed the works code to main so this PR becomes easier, I can do a similar thing for the quality control code if you want. I can't really review that code because you are the expert there. The only thing I would suggest is to add docstrings.
3) You should still use the iterator in the stimuli interface.

So, I have only two requirements to merge:
1) Decide how do you want to split the session_id as we discussed today.
2) Use the SingleImageIterator in the stimuli interface. 

As soon as we do that we can merge so you can test this in your data!

File: src/dicarlo_lab_to_nwb/thresholding_pipeline/metadata.yaml
Comment (yoonbai) [reply src/dicarlo_lab_to_nwb/thresholding_pipeline/metadata.yaml:1]: Let's keep this now and use it to register information that won't change across projects, e.g., hardware, institution, etc.

File: src/dicarlo_lab_to_nwb/conversion/psth.py
Comment (yoonbai) [reply src/dicarlo_lab_to_nwb/conversion/psth.py:105]: Nope, not intended. Will keep it

File: src/dicarlo_lab_to_nwb/conversion/stimuli_interface.py
Comment (yoonbai) [reply src/dicarlo_lab_to_nwb/conversion/stimuli_interface.py:212]: Good catch. My local changes were not committed before. Changed the following:
1. line 221: saving hash using the description field in image_kwargs
2. line 218: i encountered images with 2 channels (grayscale + transparency channel) and added an if-statement to prevent the code throwing errors

File: src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_project.py
Comment (yoonbai) [reply src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_project.py:128]: Agreed. Let's put a pin on this for now and revisit down the road, especially if we get to expand the approach for spike sorting. For spike sorting, e.g., kilosort, we might need an interface to handle a lot more user inputs

File: src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_session.py
Comment (yoonbai) [reply src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_session.py:1]: Yes. I think the whole folder, '[src/dicarlo_lab_to_nwb/thresholding_pipeline](https://github.com/catalystneuro/dicarlo-lab-to-nwb/tree/yoon_updated/src/dicarlo_lab_to_nwb/thresholding_pipeline)' could be erased after moving the following files to '[src/dicarlo_lab_to_nwb/conversion]
https://github.com/catalystneuro/dicarlo-lab-to-nwb/tree/yoon_updated/src/dicarlo_lab_to_nwb/conversion':
1. convert_project.py
2. parse_mworks_RSVP.py
3. project_config.yaml

Comment (h-mayorquin) [reply src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_session.py:1]: If you want to move that direction that's OK with me : )

File: src/dicarlo_lab_to_nwb/conversion/convert_session.py
Comment (yoonbai) [reply src/dicarlo_lab_to_nwb/conversion/convert_session.py:78]: Fixed splitting the recording_id into stimulus_name (camel case), date, and time. 
Also modified lines 314-315 in aggregation.py to follow the session_id convention written here in line 73

File: src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_project.py
Comment (yoonbai) [reply src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_project.py:154]: Agreed and updated

File: src/dicarlo_lab_to_nwb/conversion/convert_session.py
Comment (h-mayorquin) [reply src/dicarlo_lab_to_nwb/conversion/convert_session.py:78]: Thanks for doing the second :+1:

File: src/dicarlo_lab_to_nwb/thresholding_pipeline/data_locator.py
Comment (yoonbai) [reply src/dicarlo_lab_to_nwb/thresholding_pipeline/data_locator.py:1]: Not really, but will keep in case we need to reuse them later.

File: src/dicarlo_lab_to_nwb/conversion/convert_session.py
Comment (yoonbai) [reply src/dicarlo_lab_to_nwb/conversion/convert_session.py:78]: Will push updates to yoon_updated after testing

File: src/dicarlo_lab_to_nwb/conversion/stimuli_interface.py
Comment (yoonbai) [reply src/dicarlo_lab_to_nwb/conversion/stimuli_interface.py:215]: Good to know about the SingleImageIterator object. Will move forward with this implement for now until you reach a verdict with the DANDI team in using pointers, as we do with video stimuli.

File: src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_session.py
Comment (yoonbai) [reply src/dicarlo_lab_to_nwb/thresholding_pipeline/convert_session.py:1]: Great. Will do

File: src/dicarlo_lab_to_nwb/conversion/stimuli_interface.py
Comment (h-mayorquin) [inline src/dicarlo_lab_to_nwb/conversion/stimuli_interface.py:167]: Make sense.

File: src/dicarlo_lab_to_nwb/conversion/stimuli_interface.py
Comment (h-mayorquin) [inline src/dicarlo_lab_to_nwb/conversion/stimuli_interface.py:155]: Interesting, so this was failng : O
GrayScale with luminance as you were mentioning.

**Comment by h-mayorquin** [comment]
Let me know when is ready for a final check.

File: src/dicarlo_lab_to_nwb/conversion/stimuli_interface.py
Comment (yoonbai) [reply src/dicarlo_lab_to_nwb/conversion/stimuli_interface.py:155]: Yeah, this is the reason I am using the extended class and added an if-statement with the following: https://github.com/catalystneuro/dicarlo-lab-to-nwb/blob/0aa9a6edc24b83b6fb674a8065afc37003f8291c/src/dicarlo_lab_to_nwb/conversion/stimuli_interface.py#L244-L246

**Comment by yoonbai** [comment]
Ready for final check. Thanks!

**Review by h-mayorquin — APPROVED** [review]
OK.

**Pull request merged**
